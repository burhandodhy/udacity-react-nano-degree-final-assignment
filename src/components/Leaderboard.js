import { useSelector } from "react-redux";

const Leaderboard = () => {

	const users = useSelector((state) => state.users);

	const questions = useSelector((state) => state.questions);

	const usersArray = Object.values(users);
	usersArray && usersArray.sort((a, b) => (Object.keys(b.answers).length + b.questions.length) - (Object.keys(a.answers).length + a.questions.length));

	console.log(usersArray);


	return (
		<div className="container mt-5">
			<table className="table table-bordered">
				<thead className="table-light">
					<tr>
						<th style={{ width: '60%' }}>Users</th>
						<th style={{ width: '20%' }}>Answered</th>
						<th style={{ width: '20%' }}>Created</th>
					</tr>
				</thead>
				<tbody>
					{usersArray.map((user) => (
						<tr key={user.id}>
							<td>
								<img src={user.avatarURL} alt={user.name} className="avatar-img" />
								{user.name}
								<br />
								{user.id}
							</td>
							<td>{Object.keys(user.answers).length}</td>
							<td>{user.questions.length}</td>
						</tr>
					))}
				</tbody>
			</table>
		</div>
	);
};


export default Leaderboard
