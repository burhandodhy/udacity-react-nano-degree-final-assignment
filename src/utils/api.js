import { _getUsers, _getQuestions } from './_DATA';

const getInitialData = () => {
	return Promise.all([_getUsers(), _getQuestions()]).then(([users, questions]) => ({
		users,
		questions,
	}));
};




export { getInitialData }
