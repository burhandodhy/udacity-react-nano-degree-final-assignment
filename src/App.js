import { useEffect } from 'react';
import { connect, useSelector } from 'react-redux';
import { handleInitialData } from './actions/shared';

import { Container } from 'react-bootstrap';

import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';

import HeaderNavbar from './components/HeaderNavbar';
import Login from './pages/Login';
import Home from './pages/Home';
import SinglePoll from './pages/SinglePoll';
import NewPoll from './pages/NewPoll';
import Leaderboard from './pages/Leaderboard';
import NotFound from './pages/NotFound';

import LoadingBar from 'react-redux-loading-bar';

import 'bootstrap/dist/css/bootstrap.min.css';
import './App.css';

function App({ dispatch, loading }) {
	const user = useSelector((state) => state.auth);
	const isAuthenticated = user.id;

	useEffect(() => {
		dispatch(handleInitialData());
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, []);

	return (
		<>
			<LoadingBar />
			{loading === true ? null : (
				<Router>
					<HeaderNavbar />
					<Container className="mt-5">
						<Routes>
							<Route path="/" exact element={!isAuthenticated ? <Login /> : <Home />} />
							<Route path="/questions/:id" element={!isAuthenticated ? <Login /> : <SinglePoll />} />
							<Route path="/add" element={!isAuthenticated ? <Login /> : <NewPoll />} />
							<Route path="/leaderboard" element={!isAuthenticated ? <Login /> : <Leaderboard />} />
							<Route path="*" element={!isAuthenticated ? <Login /> : <NotFound />} />
						</Routes>
					</Container>
				</Router>
			)}
		</>
	);
}

const mapStateToProps = (state) => {
	return {
		loading: !(Object.keys(state.users).length > 0),
	};
};
export default connect(mapStateToProps)(App);
