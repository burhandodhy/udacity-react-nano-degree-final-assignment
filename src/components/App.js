import { useEffect } from 'react';
import { connect, useSelector } from 'react-redux';
import { handleInitialData } from './../actions/shared';

import { Container } from 'react-bootstrap';

import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';

import HeaderNavbar from './HeaderNavbar';
import Login from '../pages/Login';
import Home from '../pages/Home';
import SinglePoll from '../pages/SinglePoll';
import NewPoll from '../pages/NewPoll';
import Leaderboard from '../pages/Leaderboard';

import LoadingBar from 'react-redux-loading-bar';

import 'bootstrap/dist/css/bootstrap.min.css';
import './../App.css';

function App({ dispatch, loading }) {
	const user = useSelector((state) => state.auth);
	const isAuthenticated = user.id;

	useEffect(() => {
		dispatch(handleInitialData());
	}, []);

	return (
		<>
			<LoadingBar />
			<Router>
				<HeaderNavbar />
				<Container className="mt-5">
					{loading === true ? null : (
						<Routes>
							<Route path="/" exact element={!isAuthenticated ? <Login /> : <Home />} />
							<Route path="/questions/:id" element={!isAuthenticated ? <Login /> : <SinglePoll />} />
							<Route path="/add" element={!isAuthenticated ? <Login /> : <NewPoll />} />
							<Route path="/leaderboard" element={!isAuthenticated ? <Login /> : <Leaderboard />} />
						</Routes>
					)}
				</Container>
			</Router>
		</>
	);
}

const mapStateToProps = (state) => {
	return {
		loading: !(Object.keys(state.users).length > 0),
	};
};
export default connect(mapStateToProps)(App);
