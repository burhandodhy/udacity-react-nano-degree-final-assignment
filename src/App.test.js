import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { loadingBarReducer } from 'react-redux-loading-bar';
import { _saveQuestionAnswer, _saveQuestion } from './utils/_DATA';
import App from './App';
import Leaderboard from './pages/Leaderboard';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const qid = 'vthrdm985a262al8qx3do';
const answer = 'optionOne';
const authedUser = 'sarahedo';

const store = {
	auth: {},
	users: {
		sarahedo: {
			id: 'sarahedo',
			password: 'password',
			name: 'Sarah Edo',
			avatarURL: 'https://tylermcginnis.com/would-you-rather/sarah.jpg',
			answers: {
				'8xf0y6ziyjabvozdd253nd': 'optionOne',
				'6ni6ok3ym7mf1p33lnez': 'optionOne',
				am8ehyc8byjqgar0jgpub9: 'optionTwo',
				loxhs1bqm25b708cmbf3g: 'optionTwo',
			},
			questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9'],
		},
	},
	questions: {
		'8xf0y6ziyjabvozdd253nd': {
			id: '8xf0y6ziyjabvozdd253nd',
			author: 'sarahedo',
			timestamp: 1467166872634,
			optionOne: {
				votes: [],
				text: 'Build our new application with Javascript',
			},
			optionTwo: {
				votes: [],
				text: 'Build our new application with Typescript',
			},
		},
	},
	loadingBar: loadingBarReducer,
};

const mockedStore = mockStore(store);

test('Return Login Page', () => {
	const { container } = render(
		<Provider store={mockStore(store)}>
			<App />
		</Provider>
	);
	expect(container).toMatchSnapshot();
});

test('Renders App Name', () => {
	const { getByText } = render(
		<Provider store={mockStore(store)}>
			<App />
		</Provider>
	);

	expect(getByText('Employee Polls Web App')).toBeInTheDocument();
});

test('Renders Login Page', () => {
	const { container } = render(
		<Provider store={mockStore(store)}>
			<App />
		</Provider>
	);
	const h1Element = container.querySelector('h1');

	expect(h1Element).toBeInTheDocument();
	expect(h1Element.textContent).toBe('Login');
});

describe('_saveQuestion', () => {
	const dummyQuestion = {
		optionOneText: 'Build our new application with Javascript',
		optionTwoText: 'Build our new application with Typescript',
		author: authedUser,
	};

	it('will return correct question format', async () => {
		const results = await _saveQuestion(dummyQuestion);
		expect(results.optionOne.votes.length).toEqual(0);
		expect(results.optionTwo.votes.length).toEqual(0);
		expect(results.optionOne.text).toEqual(dummyQuestion.optionOneText);
		expect(results.optionTwo.text).toEqual(dummyQuestion.optionTwoText);
	});

	it('will return error when it not has author info', async () => {
		_saveQuestion({
			...dummyQuestion,
			author: '',
		}).catch((err) => expect(err).toEqual('Please provide optionOneText, optionTwoText, and author'));
	});
});

describe('_saveQuestionAnswer', () => {
	test('Returns updated state when correct data is passed', async () => {
		const result = await _saveQuestionAnswer({ authedUser, qid, answer });

		expect(result.questions[qid].optionOne.votes).toContain(authedUser);
		expect(result.questions[qid].optionOne.votes.length).toEqual(2);
	});

	test('Returns error when data is missing', async () => {
		try {
			await _saveQuestionAnswer({ authedUser: '', qid, answer });
		} catch (error) {
			expect(error).toBe('Please provide authedUser, qid, and answer');
		}
	});
});

test('Login form returns error when incorrect credentials add', async () => {
	const { getByTestId } = render(
		<Provider store={mockedStore}>
			<App />
		</Provider>
	);

	const usernameInput = getByTestId('username');
	const passwordInput = getByTestId('password');
	const submitButton = getByTestId('submit');

	fireEvent.change(usernameInput, { target: { value: 'sarahedo' } });
	fireEvent.change(passwordInput, { target: { value: 'incorrect' } });
	fireEvent.click(submitButton);

	await waitFor(() => {
		const actions = mockedStore.getActions();
		expect(actions).toContainEqual({ type: 'LOGIN_ERROR', payload: 'Incorrect Username or Password.' });
	});
});

test('Login user can access the dashboard', () => {
	const newStore = {
		...store,
		auth: {
			id: 'sarahedo',
		},
	};

	const newMockedStore = mockStore(newStore);

	const { getByText } = render(
		<Provider store={newMockedStore}>
			<App />
		</Provider>
	);

	const newQuestionsElement = getByText('Done');

	expect(newQuestionsElement).toBeInTheDocument();
});

test('There is logout button for logged in users', () => {
	const newStore = {
		...store,
		auth: {
			id: 'sarahedo',
		},
	};

	const newMockedStore = mockStore(newStore);

	const { getByText } = render(
		<Provider store={newMockedStore}>
			<App />
		</Provider>
	);

	expect(getByText('Logout')).toBeInTheDocument();
});

test('Nav items for logged in users', () => {
	const newStore = {
		...store,
		auth: {
			id: 'sarahedo',
		},
	};

	const newMockedStore = mockStore(newStore);

	const { getByText } = render(
		<Provider store={newMockedStore}>
			<App />
		</Provider>
	);

	expect(getByText('Home')).toBeInTheDocument();
	expect(getByText('Leaderboard')).toBeInTheDocument();
	expect(getByText('New Poll')).toBeInTheDocument();
});

test('Leaderboard page', () => {
	const newStore = {
		...store,
		auth: {
			id: 'sarahedo',
		},
	};

	const newMockedStore = mockStore(newStore);

	const { container } = render(
		<Provider store={newMockedStore}>
			<Leaderboard />
		</Provider>
	);

	expect(container).toMatchSnapshot();
});
