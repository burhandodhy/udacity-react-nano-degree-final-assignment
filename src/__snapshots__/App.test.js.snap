// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard page 1`] = `
<div>
  <div
    class="container mt-5"
  >
    <table
      class="table table-bordered"
    >
      <thead
        class="table-light"
      >
        <tr>
          <th
            style="width: 60%;"
          >
            Users
          </th>
          <th
            style="width: 20%;"
          >
            Answered
          </th>
          <th
            style="width: 20%;"
          >
            Created
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <img
              alt="Sarah Edo"
              class="avatar-img"
              src="https://tylermcginnis.com/would-you-rather/sarah.jpg"
            />
            Sarah Edo
            <br />
            sarahedo
          </td>
          <td>
            4
          </td>
          <td>
            2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Login user can access the dashboard 1`] = `
<div>
  <div />
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      Employee Polls Web App
    </a>
    <div
      class="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        class="mr-auto navbar-nav"
      >
        <a
          class="nav-link"
          href="/"
        >
          Home
        </a>
        <a
          class="nav-link"
          href="/leaderboard"
        >
          Leaderboard
        </a>
        <a
          class="nav-link"
          href="/add"
        >
          New Poll
        </a>
      </div>
      <div
        class="justify-content-end navbar-collapse collapse"
      >
        <div
          class="navbar-nav"
        >
          <div
            class="d-flex align-items-center"
          >
            <span
              class="mr-3"
            >
              Logged in as 
            </span>
            <img
              alt="User Avatar"
              class="avatar-img"
            />
            <button
              class="btn btn-outline-primary"
              type="button"
            >
              Logout
            </button>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="mt-5 container"
  >
    <div>
      <h2>
        Done
      </h2>
      <div
        class="mt-4 row"
      >
        <div
          class="mb-4 col-md-4"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <div
                class="card-title h5"
              >
                sarahedo
              </div>
              <p
                class="card-text"
              >
                6/29/2016, 7:21:12 AM
              </p>
              <a
                class="btn btn-primary"
                href="/questions/8xf0y6ziyjabvozdd253nd"
              >
                 
                Show
                 
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Return Login Page 1`] = `
<div>
  <div />
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      Employee Polls Web App
    </a>
  </nav>
  <div
    class="mt-5 container"
  >
    <div>
      <h1>
        Login
      </h1>
      <form
        class=""
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            Username
          </label>
          <input
            class="form-control"
            data-testid="username"
            placeholder="Enter username"
            type="text"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            Password
          </label>
          <input
            class="form-control"
            data-testid="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <button
          class="btn btn-primary"
          data-testid="submit"
          type="submit"
        >
          Login
        </button>
      </form>
    </div>
  </div>
</div>
`;
